airflow/api_connexion/endpoints/task_instance_endpoint.py

```diff
@@ -297,15 +297,33 @@ def post_set_task_instances_state(dag_id, session):
         error_message = f"Task ID {task_id} not found"
         raise NotFound(error_message)
 
-    execution_date = data['execution_date']
-    try:
-        session.query(TI).filter_by(execution_date=execution_date, task_id=task_id, dag_id=dag_id).one()
-    except NoResultFound:
-        raise NotFound(f"Task instance not found for task {task_id} on execution_date {execution_date}")
+    execution_date = data.get('execution_date')
+    run_id = data.get('dag_run_id')
+    if not run_id and not execution_date:
+        raise BadRequest(detail="You must provide either the dag_run_id or the execution_date")
+    if run_id and execution_date:
+        raise BadRequest(detail="You cannot provide both the dag_run_id and the execution_date")
```

Comment:
I think this can be better handled in the schema (`set_task_instance_state_form`) instead. There are existing examples for this.